feat: add streaming support for query execution

dwmkerr opened the pull request:
## Summary
- Restore streaming implementation for query controller
- Add SSE streaming support for real-time AI responses
- Include memory interface extensions for streaming

## Details
This PR restores the streaming implementation that was previously developed but removed. The code includes:
- Query controller streaming methods (`checkAndSetupStreaming`, `executeModelWithStreaming`)
- Streaming annotations (`StreamingEnabled`, `StreamingURL`)
- Memory interface streaming methods (`NotifyCompletion`, `StreamChunk`)
- HTTP streaming implementation (471 lines in `memory_http_streaming.go`)

Based on commit 5967aed4 from September 18, 2025.

**Status: DRAFT - Integration branch for streaming feature**

dwmkerr committed:
feat: devspace live reload for ark-controller

- Implement live reload for ark-controller with onUpload.restartContainer
- Update root devspace.yaml with proper ark-controller config
- Remove duplicate ark/devspace.yaml configuration
- Update .dockerignore files for ark-api and ark-dashboard
- Update local development documentation
- Standardize devspace exclusion patterns

dwmkerr committed:
feat: devspace live reload for ark-controller

- Implement live reload for ark-controller with onUpload.restartContainer
- Update root devspace.yaml with proper ark-controller config
- Remove duplicate ark/devspace.yaml configuration
- Update .dockerignore files for ark-api and ark-dashboard
- Update local development documentation
- Standardize devspace exclusion patterns

dwmkerr committed:
Merge branch 'feat/devex-ark-controller' of github.com:mckinsey/agents-at-scale-ark into feat/devex-ark-controller

dwmkerr committed:
fix: devspace controller name

dwmkerr committed:
chore: fix devspace

allcontributors[bot] committed:
docs: add kuriantom369 as a contributor for doc, and bug (#61)

Adds @kuriantom369 as a contributor for doc, bug.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/59#issuecomment-3262231982)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

kuriantom369 committed:
docs: fix broken user-guide links (#59)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #eg101 

**Description**
Fixed broken links in the `docs/user_guide` section. The issue was
caused by a base path mismatch that resulted in 404 errors. Updated the
links to use the correct base path.

**Changes**

* Corrected broken links in user guide documentation
* Verified that links now resolve to the intended pages

**Linked Issue**
Closes
[[#683](https://github.com/McK-Internal/agents-at-scale/issues/683)](https://github.com/McK-Internal/agents-at-scale/issues/683)

Signed-off-by: Tom Kurian <tom_kurian@mckinsey.com>
Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

dwmkerr committed:
chore: deploy dependencies

dwmkerr committed:
chore: fix ark controller logs

dwmkerr committed:
chore: interim; devspace dev works

dwmkerr committed:
interim: devspace deploy works

dwmkerr committed:
interim: deploy mode working

dwmkerr committed:
feat(ARKQB-227): implement ark-cluster-memory service with Memory API compliance

- Add in-memory storage service for conversation persistence
- Implement Memory API specification endpoints (POST/GET /messages, GET /sessions)
- Add DevSpace configuration for local development
- Create Memory CRD resource with setAsDefault support
- Store messages with full metadata (session_id, query_id, timestamp)
- Add required field validation for session_id and query_id

dwmkerr committed:
feat(ARKQB-227): add file-based persistence to ark-cluster-memory

- Add save/load functionality with MEMORY_FILE_PATH environment variable
- Implement automatic save on message operations
- Add periodic save every 30 seconds
- Save memory on graceful shutdown
- Add Helm chart support for PVC-based persistence
- Enable persistence in devspace configuration
- Add detailed logging for load/save operations showing message and session counts

dwmkerr committed:
feat: integrate OpenAI-compatible streaming with ark-cluster-memory

- Add Server-Sent Events streaming endpoint GET /stream/{uid}
- Add session completion endpoint POST /session/{id}/complete
- Implement OpenAI-compatible response format with finish_reason
- Add content chunking with configurable max-chunk-size parameter
- Integrate streaming with existing persistence layer
- Add comprehensive chunk logging for debugging
- Update memory API documentation with streaming endpoints
- Support session waiting and real-time event emission

Maintains full backward compatibility with existing REST API.

dwmkerr committed:
feat: add NotifyCompletion calls to ARK controller execution methods

- Add NotifyCompletion() calls to executeAgent, executeTeam, executeModel
- Enable streaming notification after query completion when memory supports streaming
- Add annotation-based streaming control via ark.mckinsey.com/memory-event-stream-enabled
- Add comprehensive logging for streaming enablement decisions
- Maintain backward compatibility with non-streaming memory services

dwmkerr committed:
feat: implement hybrid streaming architecture for real-time chunk delivery

- Add StreamChunk interface to MemoryInterface for real-time streaming
- Add persistent streaming connection support in HTTPMemory
- Add /stream-in/{sessionId} endpoint to ark-cluster-memory service
- Enable real-time chunk forwarding from ARK controller to memory service
- Add comprehensive error handling and logging for streaming operations
- Support newline-delimited JSON for efficient chunk streaming
- Maintain backward compatibility with existing memory operations

This enables real-time streaming during model execution rather than only
after completion, providing true streaming experience for end users.

dwmkerr committed:
feat: add HTTPMemory validation for streaming queries

- Add NewMemoryForQueryWithStreamingCheck for streaming capability validation
- Ensure streaming queries only use HTTPMemory (not NoopMemory)
- Provide clear error messages when streaming requirements aren't met
- Maintain backward compatibility with non-streaming queries

This prevents runtime errors when streaming is required but memory service
doesn't support it, providing early validation and clear error messages.

dwmkerr committed:
feat: add comprehensive API layer streaming support

- Add StreamingInfo model and streaming fields to QueryCreateRequest/QueryDetailResponse
- Set ark.mckinsey.com/streaming-enabled annotation when streaming=true in query creation
- Add streaming proxy endpoint GET /api/v1/namespaces/{namespace}/queries/{query_name}/stream
- Consolidate memory service endpoints: /stream-in/{sessionId} → /stream/{sessionId} (POST)
- Move completion endpoint: /session/{id}/complete → /stream/{sessionId}/complete
- Update query detail responses to include streaming URLs when enabled
- Add comprehensive streaming API documentation with ARK-specific rationale
- Handle default memory resolution following ARK controller patterns

Enables dashboard to access real-time streaming via CORS-free proxy with proper
SSE line buffering and OpenAI-compatible chunk forwarding.

dwmkerr committed:
feat: standardize streaming annotation constants across Go and TypeScript

- Add StreamingEnabled constant to ark/internal/annotations/annotations.go
- Update TypeScript constants to mirror Go structure with comments
- Fix dashboard streaming checkbox to read from metadata.annotations
- Update API to return full metadata including annotations
- Add missing general annotations (TRIGGERED_FROM, LOCALHOST_GATEWAY_PORT)

The dashboard now correctly shows streaming checkbox as checked when
viewing queries with ark.mckinsey.com/streaming-enabled annotation.

dwmkerr committed:
feat: add automatic streaming annotation for OpenAI completions API

- Add STREAMING_ENABLED_ANNOTATION constant to ark-api annotations
- Update /openai/v1/chat/completions endpoint to set streaming annotation
  when request.stream=true
- Use shared annotation constants instead of hardcoded strings

When clients call the OpenAI-compatible API with stream=true, queries
are automatically created with the ark.mckinsey.com/streaming-enabled
annotation, enabling real-time streaming responses.

dwmkerr committed:
feat: enable streaming support in ark-cluster-memory helm chart

- Add ark.mckinsey.com/memory-event-stream-enabled annotation to memory resource
- Include clear comments explaining streaming functionality
- Enables OpenAI completions API streaming when this memory is used

This annotation tells the ARK API that this memory service supports
Server-Sent Events streaming for real-time query response updates.

dwmkerr committed:
feat: add enableStreaming parameter to ark-cluster-memory chart

- Add memory.enableStreaming parameter in values.yaml (default: true)
- Make streaming annotation conditional based on chart parameter
- Clear comments distinguish between streaming support vs cluster enablement

This allows operators to disable streaming in specific cluster deployments
while the memory service still technically supports streaming functionality.

dwmkerr committed:
feat: update wait-for-session parameter format in memory service

- Update ark-cluster-memory to support wait-for-session=<timeout> format
- Maintains backward compatibility with wait-for-session=true
- Update memory reference docs to reflect correct parameter format

Example: wait-for-session=30s instead of separate timeout parameter

dwmkerr committed:
fix: properly terminate SSE streams in memory service

- Add res.end() call after sending data: [DONE] marker
- Ensures HTTP connection closes cleanly after stream completion
- Fixes timeout issues in OpenAI streaming API and clients

This resolves the issue where SSE streams would hang indefinitely
waiting for connection termination after content was complete.

dwmkerr committed:
fix: use proper OpenAI completion message format in memory service

- Change completion message from {type: 'completion', finish_reason: 'stop'}
  to {choices: [{finish_reason: 'stop', delta: {}}]}
- Matches OpenAI streaming specification for completion chunks
- Fixes SSE stream termination when ARK controller calls NotifyCompletion

This ensures the SSE handler properly detects completion and sends
'data: [DONE]' followed by connection close.

dwmkerr committed:
feat: implement OpenAI-compatible streaming with annotation-based control

- Add streaming support to ChatCompletionProvider interface with n parameter and ChatCompletionStream method
- Update all providers (OpenAI, Azure, Bedrock) to implement streaming interface
- Implement streaming logic in model.ChatCompletion() with memory.StreamChunk() calls
- Add streaming annotation detection in query controller
- Fix memory service to emit streaming chunks without persisting them
- Add separate chunk: and message: event types for proper streaming vs persistence
- Update method signatures across codebase to support streaming parameters
- Streaming works for model-level queries with ark.mckinsey.com/streaming-enabled annotation

dwmkerr committed:
fix: update memory service to separate streaming chunks from persistent messages

- Change streaming chunks to emit chunk: events instead of message: events
- Make eventEmitter public to allow direct chunk emission
- Add subscribeToChunks() method for streaming clients
- Streaming chunks no longer get persisted as individual messages
- Only final messages via /messages endpoint get saved to storage

dwmkerr committed:
docs: clarify streaming API uses query execution context not conversation sessions

- Change {session_id} to {query_id} in streaming endpoint documentation
- Clarify that streaming is per-query execution, not long-lived sessions
- Update parameter names: wait-for-session -> wait-for-query
- Each query gets its own streaming context identified by query UID
- Streaming endpoints handle individual query execution lifecycle

dwmkerr committed:
feat: use query name for streaming endpoints instead of query UID

- Add queryName field to HTTPMemory struct alongside sessionId
- Update streaming endpoints to use query name: POST /stream/{query_name}
- Add QueryName to Config struct and pass from controller
- Streaming now uses human-readable query names instead of UUIDs
- /messages endpoint still uses both session_id (query UID) and query_id (query name)
- Fix stream.Close() error handling in providers

dwmkerr committed:
feat: update memory service streaming endpoints to use query_id

- Change /stream/{sessionId} to /stream/{query_id} for streaming endpoints
- Update completion endpoint to /stream/{query_id}/complete
- Rename completeSession to completeQueryStream for clarity
- Remove sessionExists check from completion (streams vs messages)
- Emit completion markers to chunk: events instead of message: events
- Update all logging to use 'query' terminology instead of 'session'

dwmkerr committed:
feat: implement proper streaming separation and chunk storage in memory service

- Separate streaming routes from memory routes for better organization
- Store streaming chunks in dedicated Map separate from messages
- Fix issue where reconnecting to completed streams would hang
- Add consistent logging every 10 chunks for both inbound and outbound
- Remove verbose per-chunk logging from ark controller
- Properly close connections for already-completed streams

dwmkerr committed:
fix: use aiter_lines() for true streaming in OpenAI API

The previous implementation using aiter_text() was buffering the entire response.
Using aiter_lines() provides true line-by-line streaming without buffering,
allowing clients to receive chunks as they are generated.

paliwalvimal committed:
docs: update repo url from mck-internal to mckinsey for clone instruction (#67)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [x] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #eg101 


closes #68

allcontributors[bot] committed:
docs: add paliwalvimal as a contributor for doc, and bug (#70)

Adds @paliwalvimal as a contributor for doc, bug.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/67#issuecomment-3265172870)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

Nandika2000 committed:
feat: agents as tools (#40)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

---------

Co-authored-by: Nandika.Arora.ext <Nandika.Arora.ext@daiichi-sankyo.eu>
Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

dwmkerr committed:
feat: implement transparent OpenAI chunk streaming with proper completion signaling

- Forward all OpenAI chunks as-is (text, tool calls, finish_reason) without filtering
- Use explicit completion event instead of finish_reason for stream termination
- Support multi-agent scenarios with multiple intermediate finish_reason chunks
- Add comprehensive Jest tests for streaming endpoints with NDJSON format
- Create streaming documentation in developer guide
- Update memory service to handle completion events separately from chunks

dwmkerr committed:
fix: use separate HTTP client without timeout for streaming connections

- Create dedicated HTTP client for streaming to avoid 30-second timeout on long-lived connections
- Update error messages to clearly distinguish ARK controller vs client disconnections
- Successfully tested with 900+ chunk streams

dwmkerr committed:
docs: add disconnect behavior notes for streaming APIs

dwmkerr committed:
feat: enable streaming for agents

- Pass eventStream (MemoryInterface) and streamingEnabled flag through agent execution
- Update TeamMember interface to support streaming parameters
- Agents now stream chunks in real-time just like direct model calls
- Teams pass nil/false as streaming is not yet supported for teams
- Successfully tested with 700+ chunk streams without timeout issues

anshu-rani-singh committed:
feat: agent as tool creation (#43)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101 

NOTE: This PR should be merged after merging this PR
https://github.com/mckinsey/agents-at-scale-ark/pull/40

Screenshots:
<img width="672" height="770" alt="image"
src="https://github.com/user-attachments/assets/20db8165-33c8-4cf3-a313-16c67f14b078"
/>

<img width="1715" height="866" alt="image"
src="https://github.com/user-attachments/assets/3cd7c29d-a417-46cb-ab5b-8ea6ee608a1c"
/>

<img width="669" height="880" alt="image"
src="https://github.com/user-attachments/assets/8a965aca-988b-4d00-83c8-0c70c9881826"
/>

Recording:


https://github.com/user-attachments/assets/c27532cb-ec68-4253-85b7-169c693d701a

---------

Co-authored-by: Nandika.Arora.ext <Nandika.Arora.ext@daiichi-sankyo.eu>

allcontributors[bot] committed:
docs: add anshu-rani-singh as a contributor for code (#83)

Adds @anshu-rani-singh as a contributor for code.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/43#issuecomment-3272395070)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
docs: add havasik as a contributor for code, and bug (#84)

Adds @havasik as a contributor for code, bug.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/73#issuecomment-3272405620)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

allcontributors[bot] committed:
docs: add jskeates as a contributor for bug (#81)

Adds @jskeates as a contributor for bug.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/80#issuecomment-3272106427)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>
Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

jskeates committed:
feat(dashboard): Define config as map in Helm chart values (#80)

This is needed to make the values easily override-able. For example, to
set the host of ark-api (if running in a non-default namespace), you can
now set `app.config.arkApiService.host`, rather than fragile array
overrides like `.app.env[3].value`.

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [x] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

dwmkerr committed:
feat: update charts to use GHCR images by default (#86)

- Rename `ark` chart to `ark-controller`
- Build/deploy ark api / dashboard / mcp charts
- Default to GHCR images for production use
- Local development uses local images

skokaina committed:
ci: fix lower strictness for tool evals (#91)

## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

github-actions[bot] committed:
chore(main): release 0.1.33 (#85)

:robot: I have created a release *beep* *boop*
---


##
[0.1.33](https://github.com/mckinsey/agents-at-scale-ark/compare/v0.1.32...v0.1.33)
(2025-09-10)


### Features

* agent as tool creation
([#43](https://github.com/mckinsey/agents-at-scale-ark/issues/43))
([4b58aa3](https://github.com/mckinsey/agents-at-scale-ark/commit/4b58aa368c4cc3b8e13c887879c80b24e278196a))
* agents as tools
([#40](https://github.com/mckinsey/agents-at-scale-ark/issues/40))
([d75c1cb](https://github.com/mckinsey/agents-at-scale-ark/commit/d75c1cbe294917b0a6d51a87db84109bda52d6a3))
* **dashboard:** Define config as map in Helm chart values
([#80](https://github.com/mckinsey/agents-at-scale-ark/issues/80))
([f946aa2](https://github.com/mckinsey/agents-at-scale-ark/commit/f946aa259b420df1860712a3086fe8bf12b9e4c3))
* devspace live reload for ark-controller
([#60](https://github.com/mckinsey/agents-at-scale-ark/issues/60))
([5ac7996](https://github.com/mckinsey/agents-at-scale-ark/commit/5ac79963de8393d31ec8396005794bbcbcfda798))
* update charts to use GHCR images by default
([#86](https://github.com/mckinsey/agents-at-scale-ark/issues/86))
([fabfd38](https://github.com/mckinsey/agents-at-scale-ark/commit/fabfd38a2b544eefd1cd511f2b71ab5e2b810da0))

---
This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dwmkerr committed:
build: never cancel 'main' build due to concurrency (#92)

poornimanagQB committed:
fix: Fix Namespace and path (#100)

Bug Fix:
- Fixed webhook service namespapce to ark-system
- Fixed wheel file path matching with the path in Docker file 
## Checklist

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

anshu-rani-singh committed:
feat(dashboard): Delete confirmation modal for agent, team and tool (#90)

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #eg101 

Jira: https://mckinsey.atlassian.net/browse/ARKQB-237

Screenshot:
<img width="1473" height="962" alt="Screenshot 2025-09-10 at 3 51 13 PM"
src="https://github.com/user-attachments/assets/da9b93ec-9db1-41eb-8705-9f960cd647fd"
/>

<img width="565" height="288" alt="Screenshot 2025-09-10 at 5 13 03 PM"
src="https://github.com/user-attachments/assets/5e98e95c-333a-47e0-bf18-956a2d12daa0"
/>

<img width="516" height="244" alt="Screenshot 2025-09-10 at 5 13 38 PM"
src="https://github.com/user-attachments/assets/6c83f9f8-195d-4109-acd7-31dff5e0e182"
/>

av3n93rz committed:
feat(dashboard): adds ODIC with 'sso' and 'open' authentication models for dashboard

poornimanagQB committed:
feat: query response format (#82)

In this PR:
- Added raw JSON response view with copy/download functionality and
removed markdown view mode
- Error Handling: Created separate Error section with Events and Details
view modes, implementing error fetching from Kubernetes events with
fallback to generic messages
- Renamed "Text" to "Content" and "JSON" to "Raw" for better clarity,
restructured query detail page with conditional Response/Error section
display
- Events link has been added.

Scenario 1: When there is no error:
- content view
<img width="1705" height="906" alt="image"
src="https://github.com/user-attachments/assets/611a4b07-b5e0-49bf-becd-250f16291e8e"
/>

- raw view
<img width="1705" height="906" alt="image"
src="https://github.com/user-attachments/assets/45f31ebe-ad19-49da-8233-11a1b7122a24"
/>

Scenario 2: When there is error:
- events view
<img width="1705" height="938" alt="image"
src="https://github.com/user-attachments/assets/1132e605-58a1-4978-bc7b-57a56088bb61"
/>
- Details view
<img width="1705" height="948" alt="image"
src="https://github.com/user-attachments/assets/7881b7f1-a0a2-4ed3-9216-ce8660908efa"
/>

---------

Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

havasik committed:
fix(ark-api): return default model if not set for agent (#73)

Relates to #36

Agents without model ref are now displayed with "default" - this is also
reflected on the Models dashboard "Used by n agents.

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #eg101

Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

adriander committed:
feat: non-blocking agent creation with deferred dependency validation (#89)

ARK's admission webhook currently blocks agent creation when tool
dependencies don't exist, breaking normal Kubernetes patterns. Unlike
Secrets/ConfigMaps, agents cannot reference non-existent tools, forcing
artificial ordering constraints and preventing natural reconciliation.

Adopt the Kubernetes model for tool dependencies:
- **Non-blocking creation**: Agents create successfully even with
missing tool dependencies
- **Runtime status management**: Agents show `Pending` status when
dependencies are missing
- **Event-driven resolution**: Agents transition to `Running`
immediately when dependencies are created
- **Automatic recovery**: No manual intervention required

- **Agent Controller**: Added Tool/Model watches for immediate
dependency resolution
- **CRD Schema**: Fixed `status.phase` field with proper enum validation
and "Pending" default
- **Event-driven architecture**: Blocking agent creation with instant
status transitions
- **Deduplication**: Prevent race conditions during concurrent
reconciliations
- **Code cleanup**: Refactored duplicate dependency finder logic

Run `./test-nonblocking-agent-status.sh` to verify the event-driven
dependency resolution works correctly.

- [ ] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues
[ARKQB-197](https://mckinsey.atlassian.net/browse/ARKQB-197)

---------

Co-authored-by: Dave Kerr <dwmkerr@gmail.com>

skokaina committed:
feat: ark evaluator with langfuse (#65)

Major Features:
  - Langfuse + RAGAS integration in ark-evaluator
  - Introduce ark-evaluator as a single holistic evaluation service
  - update evaluator to handle context in langfuse adatper
  - refactor tests

  Bug Fixes:
  - CI/CD reliability improvements
  - uv loop threading issue with next_asyncio for Ragas
  - Tool caching and navigation fixes
  - Documentation link fixes
  - Helm chart deployment issues

  Infrastructure:
  - Enhanced documentation and UI improvements

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [x] Update `./docs`
- [x] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #eg101

---------

Co-authored-by: iamhassaan <muhammad_anwar@mckinsey.com>
Co-authored-by: Nishanth Prajith <nishanth.prajith>

dwmkerr committed:
fix: helm charts use AppVersion for image tags and deploy workflow supports latest (#95)

## Summary
- Remove hardcoded `tag: latest` from all Helm chart values.yaml files
- Update deployment templates to default to `.Chart.AppVersion` when tag
not specified
- Fix localhost-gateway appVersion to use our version instead of nginx
version
- Add localhost-gateway appVersion to release-please configuration
- Add `deploy_containers_to_latest` parameter to deploy workflow
(defaults to false)
- Update cicd workflow to push to "latest" tags when releasing
- Fix deploy_helm_chart description to use plural 'charts'

## Impact
- Helm charts now automatically use the correct versioned images based
on Chart.AppVersion
- No more manual tag updates needed in values.yaml files for releases
- Release Please only needs to update Chart.yaml files
- Production releases automatically push both versioned and "latest"
tags
- Manual deployments don't push to "latest" by default (safer)

---------

Co-authored-by: Sallah Kokaina <sallah.kokaina@gmail.com>

adriander committed:
fix: ark-api container restart in devspace (#102)

Updates wheel file path from out/ark_sdk-*.whl to /app/out/ark_sdk-*.whl to resolve self-dependency error causing infinite restart cycles during local development

AshAnand34 committed:
docs: license link in contributing.md fixed (#96)

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #88

AshAnand34 committed:
fix: update badge template URL and improve iframe usage for contributors (#98)

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [ ] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [x] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [x] Linked issues #62

allcontributors[bot] committed:
docs: add AshAnand34 as a contributor for bug, and doc (#106)

Adds @AshAnand34 as a contributor for bug, doc.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/98#issuecomment-3285519133)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

havasik committed:
fix: `devspace dev` dashboard console errors (#105)

## Description

When launching `devspace dev`, dashboard is failing to handle auth
session API request calls and logging some errors:

<img width="939" height="813" alt="session"
src="https://github.com/user-attachments/assets/9f1876cf-df27-4f57-8752-eaef7971fd40"
/>

Adding the correct NODE_ENV variable fixes it. 

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [ ] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [ ] Update `./docs`
- [ ] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101

allcontributors[bot] committed:
docs: add AshAnand34 as a contributor for doc, bug, and code (#107)

Adds @AshAnand34 as a contributor for doc, bug, code.

This was requested by dwmkerr [in this
comment](https://github.com/mckinsey/agents-at-scale-ark/pull/97#issuecomment-3285534441)

[skip ci]

---------

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

dwmkerr committed:
fix: improve streaming reliability and fix context cancellation issues

- Use context.Background() for streaming HTTP requests to prevent "context canceled" errors
- Implement dual HTTP client pattern: memoryHttpClient (with timeout) and streamHttpClient (no timeout)
- Make memory timeout configurable via ARK_MEMORY_HTTP_TIMEOUT environment variable
- Fix bug where completed streams didn't send [DONE] marker when replaying from beginning
- Update ark-cluster-memory logging to be time-based (every second) instead of chunk-based
- Add chunk type counting in logs (content, tool_calls, finish_reason, other)

These changes fix critical streaming issues where connections would timeout or show
spurious "context canceled" errors after query completion.

dwmkerr committed:
fix: preserve full message structure in streaming model execution

- Update executeModelWithStreaming to use choice.Message.ToParam() instead of just extracting content
- This preserves tool calls if present in the response (matching non-streaming behavior)
- Models don't execute tools (consistent with current architecture where tools are agent-level)

dwmkerr committed:
fix: add logging to debug tool call accumulation during streaming

dwmkerr committed:
fix: enable tool call accumulation for Azure provider during streaming

- Updated Azure provider to use same accumulateStreamChunk function as OpenAI
- Tool calls now properly accumulate from fragmented streaming chunks
- Agents can now execute tools when streaming is enabled with Azure models
- Added test script to verify streaming with tool calls works correctly
- Updated docs to clarify model support (OpenAI ✅, Azure ✅, Bedrock ❌)

dwmkerr committed:
feat: add execution metadata context keys for streaming

- Add context keys for target, team, agent, and model tracking
- Add WithExecutionMetadata and GetExecutionMetadata functions
- Prepare foundation for adding ark metadata to streaming chunks
- Keep team streaming disabled for now pending metadata implementation

dwmkerr committed:
feat: add ark metadata to streaming chunks

- Implement chunk wrapping with metadata using struct embedding
- Add wrapChunkWithMetadata function to enrich chunks with context
- Successfully adds ark field to streaming chunks with model name
- Metadata is backward compatible - doesn't break existing streams

dwmkerr committed:
feat: enrich streaming chunks with full execution metadata

- Add query ID and target to all streaming chunks
- Add agent name when executing through agents
- Direct model queries show target as model/name
- Agent queries show target as agent/name plus agent field
- Metadata helps clients differentiate execution paths

dwmkerr committed:
feat: enable streaming support for team queries

- Pass memory interface and streaming flag to team.Execute()
- Teams now support streaming when query has streaming annotations
- Matches the same streaming behavior as agent queries

dwmkerr committed:
feat: add team and agent metadata to streaming context

- Enrich context with team name and current agent when teams execute members
- Streaming chunks will now include which agent is responding within a team
- Helps identify the source of responses in multi-agent team scenarios

av3n93rz committed:
fix: missing QueryClientProvider issue, queries tab ui glitch (#108)

Fixes:
- the missing QueryClientProvider issue in the ark-dashboard
- the queries tab having 2 versions overlapping each other on the
ark-dashboard

## Checklist

- [x] Follow the [contributor guide](./CONTRIBUTING.md)
- [x] End-to-end tests pass
- [x] Unit tests for essential parts of your code, e.g. APIs exposed via
SDKs or endpoints
- [x] Update `./docs`
- [x] Recognize contributors with "@all-contributors please add <person>
for <code, bug, docs, etc>"
- [ ] Linked issues #eg101 

<img width="1920" height="1007" alt="Screenshot 2025-09-12 at 17 17 53"
src="https://github.com/user-attachments/assets/3a71d8b0-71d5-4929-a847-2840b513d3bf"
/>

adriander committed:
feat: Update agent UI to show status (#104)

Add agent status badges to dashboard UI

Adds visual status indicators (Ready/Pending/Error) to agent cards and
table rows, matching the existing model status display pattern. Includes
color-coded badges for quick agent health identification.

dwmkerr committed:
feat: add BaseURL support for Bedrock models and fix memory endpoints

- Add BaseURL field to BedrockModelConfig for AI Gateway support
- Update Bedrock provider to use custom endpoint when BaseURL provided
- Fix memory backend endpoint conflicts (messages vs memory-status, stream vs stream-statistics)
- Add comprehensive Models documentation page
- Remove Models section from CRDs page to avoid duplication
- Update streaming documentation with Bedrock status

dwmkerr committed:
feat: implement single-chunk streaming fallback for Bedrock provider

Per OpenAI spec, when streaming is requested for a model that doesn't support it,
the backend should send the complete response as a single chunk. This allows
Bedrock models to work with streaming queries.

dwmkerr committed:
fix: resolve variable shadowing issue in context_utils.go

Fixed compilation error where ctx variable was being shadowed with :=
instead of using = for assignment

dwmkerr committed:
docs: update streaming documentation to reflect Bedrock fallback behavior

Updated streaming.mdx to document that Bedrock models now return complete
responses as single chunks per OpenAI spec when streaming is requested

dwmkerr committed:
fix: allow ensure-ark-sdk-wheel.sh to work with forked repositories

dwmkerr committed:
fix: integrate agents-as-tools with streaming - update Execute signature

dwmkerr committed:
feat: add A2AServer Ready condition checking for agent dependencies

- Agents owned by A2AServers now check if the A2AServer is Ready
- Added event-driven watches for A2AServer status changes
- Agents show Pending phase when A2AServer is not ready
- Added diagnostic events: A2AServerNotFound, A2AServerNotReady
- Follows same pattern as Model and Tool dependency checking

dwmkerr committed:
feat: implement A2AServer dependency checking for agents

- Add checkA2AServerDependency function to validate A2AServer Ready condition
- Watch A2AServer events and trigger agent reconciliation when status changes
- Set agent phase to Pending when owning A2AServer is not Ready
- Fix A2AServer controller to not delete agents on discovery failure
- Follow Kubernetes pattern: resources persist, status reflects reality

dwmkerr committed:
feat: implement A2AServer dependency checking for agents

- Add checkA2AServerDependency function to validate A2AServer Ready condition
- Watch A2AServer events and trigger agent reconciliation when status changes
- Set agent phase to Pending when owning A2AServer is not Ready
- Fix A2AServer controller to not delete agents on discovery failure
- Follow Kubernetes pattern: resources persist, status reflects reality

dwmkerr committed:
feat: add A2AServerReady event when dependency is satisfied

- Emit Normal event when A2AServer becomes ready for better observability
- Complements the Warning event when A2AServer is not ready

dwmkerr committed:
feat: add A2AServerReady event when dependency is satisfied

- Emit Normal event when A2AServer becomes ready for better observability
- Complements the Warning event when A2AServer is not ready

dwmkerr committed:
merge: integrate feat/ARKQB-227-in-cluster-memory streaming and memory features

- Resolved conflicts in agent_controller.go to include A2AServer dependency checking
- Integrated streaming support with agent_tools.go Execute function
- Added ark-cluster-memory dependency to devspace.yaml
- Fixed dashboard imports to include Checkbox and streaming annotations
- Updated documentation to include Models section
- Preserved all streaming/memory features while incorporating upstream changes

dwmkerr committed:
fix: remove orphaned code from merge conflict resolution in agent_controller

dwmkerr committed:
fix: resolve linting issues in A2AServer dependency implementation

- Reduce cognitive complexity by extracting helper functions
- Fix gofumpt formatting issues
- Simplify nested conditionals
- Remove unused deleteAgentByA2AServer function
- Add sample YAML for A2AServer-Agent dependency

dwmkerr committed:
merge: integrate feat/a2aserver-agent-dependencies fixes

- Refactored A2AServer dependency checking with helper methods
- Fixed duplicate function definitions in agent_controller
- Improved code organization with validateA2AServerDependency and isA2AServerReady methods

dwmkerr committed:
docs: add streaming and memory documentation

- Added streaming query documentation in developer guide
- Added models resource documentation page
- Updated memory resource docs with cluster memory details
- Added references to streaming endpoints in API docs
- Updated resource metadata for proper navigation

dwmkerr committed:
docs: remove SSE Streaming section from ark-apis.mdx

dwmkerr committed:
docs: keep ark-apis.mdx closer to upstream with minimal streaming changes

dwmkerr committed:
docs: remove Agents section from crds.mdx and restore Model reference

dwmkerr committed:
Merge branch 'feat/ARKQB-138-streaming-docs' into feat/streaming-memory-integration

dwmkerr committed:
fix: add missing Checkbox import in query page

dwmkerr committed:
Merge branch 'upstream/main' into feat/streaming-memory-integration

Incorporates:
- A2AServer dependency checking for agents (#121)
- Container registry fixes for external PRs (#120)

dwmkerr committed:
Merge branch 'upstream/main' into feat/streaming-memory-integration

Incorporates:
- Bedrock BaseURL support (#124)
- Previous A2AServer and registry fixes

No conflicts - BaseURL changes integrate smoothly with streaming implementation

dwmkerr committed:
chore: merge upstream main and resolve memory.mdx conflicts

dwmkerr committed:
chore: remove unused annotation constants from dashboard

dwmkerr committed:
Merge branch 'main' of github.com:mckinsey/agents-at-scale-ark into feat/streaming-memory-integration

dwmkerr committed:
chore: regenerate CRDs with streaming support

dwmkerr committed:
chore: remove temporary documentation files

dwmkerr committed:
fix: address lint warnings for code quality

dwmkerr committed:
merge: integrate upstream main with streaming implementation

- Accept upstream ark-cluster-memory service implementation
- Use upstream ark-api streaming with check_streaming_availability()
- Keep streaming annotations without 'upcoming feature' comment
- Use ARK_MEMORY_HTTP_TIMEOUT_SECONDS env var from upstream
- Fix model controller probe context variable
- Add query name parameter to NewMemoryForQuery from upstream

dwmkerr committed:
fix: prevent model controller reconciliation loop

- Add stable error messages for model probe failures
- Remove request-specific details (RequestIDs) from status conditions
- Change probe failures from ERROR to INFO logs (expected state)
- Remove OperationTracker to prevent unnecessary event spam
- Only emit events for actual probe results, not routine operations
- Simplify controller by removing unnecessary processModel wrapper

The model controller now:
- Respects poll intervals properly
- Logs probe failures at INFO level with stable messages
- Emits clean events (ModelProbeFailed/ModelProbeSucceeded)
- Prevents tight reconciliation loops caused by changing error messages

dwmkerr committed:
docs: add logging and event guidelines for stable messages

- Document that event/condition messages must be strings, not JSON
- Explain why stable messages prevent reconciliation loops
- Add examples showing how RequestIDs cause tight loops
- Clarify that expected failures should be logged as INFO
- Show clean event output with count aggregation (x9 over 10m)

dwmkerr committed:
chore: clean up documentation and remove notes file

- Remove duplicate Agents section from crds.mdx (already covered in agent.mdx)
- Delete temporary notes file from ark-dashboard

dwmkerr committed:
fix: remove unnecessary streaming field from QueryCreateRequest, use metadata instead

dwmkerr committed:
chore: remove unused StreamingInfo class and stream endpoint from ark-api queries

dwmkerr committed:
refactor: separate event streaming from memory operations

- Create EventStreamInterface in ark/internal/genai/streaming.go
- Add streaming_config.py to ark_sdk for ConfigMap-based streaming discovery
- Update OpenAI completions API to use ConfigMap instead of Memory annotations
- Remove old memory-based streaming checks
- Update ark-cluster-memory Helm chart with streaming ConfigMap
- Add comprehensive tests for streaming configuration

dwmkerr committed:
fix: add configmap permissions to ark-api service account for streaming config

dwmkerr committed:
fix: use port name instead of number in streaming config

- Update ConfigMap template to use port name 'http'
- Clarify documentation that port name must be used
- ResolveServiceReference only matches port names, not numbers

dwmkerr committed:
fix: resolve port names to numbers in streaming config

- Update streaming_config.py to resolve service port names to numbers
- Change get_streaming_url to async get_streaming_base_url function
- Add Kubernetes service lookup to find actual port numbers
- Update tests to mock async service lookups
- Update ark-api to use new async base URL function
- Clean up Helm chart ConfigMap comments

dwmkerr committed:
fix: add RBAC permissions for streaming configuration

- Split configmaps and services permissions with clear comments
- Add permission to read configmaps for ark-config-streaming
- Add permission to read services for streaming service discovery

dwmkerr committed:
fix: resolve Go linting issues in ark controller

- Add proper error handling for Close operations in streaming
- Log errors for cleanup operations that fail
- Fix unused variable declaration in context_utils
- Remove unused checkAndSetupStreaming function
- Format code with gofmt
- Add nolint directive for intentional context accumulation

Reduced linting issues from 21 to 5 (remaining are complex refactoring tasks)

dwmkerr committed:
fix: resolve remaining linting issues

- Mark unused memory parameter with underscore in executeLocally
- Refactor nested if complexity with early return pattern
- Format code with gofmt

Reduced linting issues from 5 to 3 (remaining are cognitive complexity in test/provider functions)

dwmkerr committed:
fix: refactor streaming test to reduce cognitive complexity

- Extract helper functions for creating test ConfigMaps and verifying results
- Simplify test case definitions using helpers
- Fix test data to use port names (http) instead of numbers (80, 8080)
- Reduce cognitive complexity from 24 to acceptable level

Tests still pass and linting issues reduced from 3 to 2

dwmkerr committed:
fix: eliminate all Go linting issues

- Extract helper methods to reduce cognitive complexity
- Add prepareStreamParams to prepare chat completion parameters
- Add processToolCalls to handle tool call accumulation logic
- Add sendFinalToolCallChunk for sending final accumulated messages
- Apply same pattern to both OpenAI and Azure providers

All linting issues resolved (0 issues remaining from original 21)
All tests pass

dwmkerr committed:
fix: merge upstream main with streaming improvements

- Integrated upstream processModel refactoring
- Kept ProbeResult return type for better error handling
- Combined upstream's status change detection with our ProbeResult pattern
- Resolved conflicts in model_controller.go

dwmkerr committed:
fix: improve CI/CD reliability and container registry configuration

- Increase chainsaw test timeouts for LLM operations
- Fix container registry paths to include repository name for GHCR access control
- Add NPM package metadata for proper display on npmjs.com
- Fix deploy workflow parameter naming
- Separate streaming configuration from memory operations
- Implement proper port name resolution in Python SDK
- Fix Go linting issues (21 total) with proper error handling
- Merge upstream model/agent improvements with eventual consistency

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -191,13 +192,23 @@ func (r *QueryReconciler) executeQueryAsync(opCtx context.Context, obj arkv1alph
 		return
 	}
 
+	// Store eventStream for cleanup later
```
This should be clearer - if an eventstram client is used then close at the end of query resolution

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -227,6 +238,13 @@ func (r *QueryReconciler) executeQueryAsync(opCtx context.Context, obj arkv1alph
 	} else {
+		// Notify event stream that streaming is complete (if streaming was enabled)
+		if eventStream != nil {
+			if completionErr := eventStream.NotifyCompletion(opCtx); completionErr != nil {
```
could we potentially close the stream here, rather than with the defer block in line 203

dwmkerr replied:
Also if we cannot close the event stream i don't think that callers will get a successful query back is that right? if so should we fail the query?

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -382,10 +400,32 @@ func (r *QueryReconciler) resolveEvaluatorSelector(ctx context.Context, selector
-func (r *QueryReconciler) reconcileQueue(ctx context.Context, query arkv1alpha1.Query, impersonatedClient client.Client, memory genai.MemoryInterface, tokenCollector *genai.TokenUsageCollector) ([]arkv1alpha1.Response, error) {
+func (r *QueryReconciler) reconcileQueue(ctx context.Context, query arkv1alpha1.Query, impersonatedClient client.Client, memory genai.MemoryInterface, tokenCollector *genai.TokenUsageCollector) ([]arkv1alpha1.Response, genai.EventStreamInterface, error) {
+	// Create event stream if streaming is requested
+	var eventStream genai.EventStreamInterface
```
would this be better in `genai/query.go` or something as a reusable function to check if streaming was requested, with its own (simple) unit test

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -382,10 +400,32 @@ func (r *QueryReconciler) resolveEvaluatorSelector(ctx context.Context, selector
 }
 
-func (r *QueryReconciler) reconcileQueue(ctx context.Context, query arkv1alpha1.Query, impersonatedClient client.Client, memory genai.MemoryInterface, tokenCollector *genai.TokenUsageCollector) ([]arkv1alpha1.Response, error) {
+func (r *QueryReconciler) reconcileQueue(ctx context.Context, query arkv1alpha1.Query, impersonatedClient client.Client, memory genai.MemoryInterface, tokenCollector *genai.TokenUsageCollector) ([]arkv1alpha1.Response, genai.EventStreamInterface, error) {
```
would the query reconciler be better having an eventstream as a member?

dwmkerr replied:
or in context?

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -486,6 +526,17 @@ func (r *QueryReconciler) executeTarget(ctx context.Context, query arkv1alpha1.Q
 	)
 	defer span.End()
 
+	// Add query and session context for streaming metadata
```
isn't the session id already present in context at this point? on admission if we don't have a session id we generate one I beliee

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -546,6 +597,11 @@ func (r *QueryReconciler) executeAgent(ctx context.Context, query arkv1alpha1.Qu
 	log := logf.FromContext(ctx)
 	log.Info("executing agent", "agent", agentCRD.Name)
 
+	// Add agent to execution metadata
```
+ this ensures that clients can see the specific agent being queried when streaming, as per the 'streaming' docs

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -650,33 +706,44 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
+			return nil, err
+		}
+	} else {
+		// Execute without streaming (existing logic)
```
no need for breadcrumbs

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -650,33 +706,44 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
+	if eventStream != nil {
+		// Execute with streaming
+		var err error
+		responseMessages, err = r.executeModelWithStreaming(ctx, model, allMessages, eventStream, modelTracker)
```
how do we get token usage here?

dwmkerr replied:
in facy we get it in the execute model with streaming but make this clear with a comment

dwmkerr commented on ark/internal/controller/query_controller.go:
```diff
@@ -650,33 +706,44 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
+			CompletionTokens: completion.Usage.CompletionTokens,
+			TotalTokens:      completion.Usage.TotalTokens,
+		}
+		modelTracker.CompleteWithTokens("", tokenUsage)
```
what is the empty string here

dwmkerr commented on ark/internal/genai/agent.go:
```diff
@@ -227,13 +229,33 @@ func (a *Agent) executeLocally(ctx context.Context, userInput Message, history [
 		newMessages = append(newMessages, assistantMessage)
 
+		// Debug logging for streaming tool calls
+		logger := logf.FromContext(ctx)
```
we don't need debug logging here

dwmkerr commented on ark/internal/genai/agent.go:
```diff
@@ -227,13 +229,33 @@ func (a *Agent) executeLocally(ctx context.Context, userInput Message, history [
+		}
+
 		if len(choice.Message.ToolCalls) == 0 {
+			logger.Info("No tool calls, returning", "agent", a.FullName())
```
same, the logging on this line and 253 and 255 is superfluous

dwmkerr commented on ark/internal/genai/agent_tools.go:
```diff
@@ -280,7 +280,9 @@ func (a *AgentToolExecutor) Execute(ctx context.Context, call ToolCall, recorder
 
 	// Call the agent's Execute function
-	responseMessages, err := agent.Execute(ctx, userInput, history)
+	// Pass nil for memory and eventStream (agents-as-tools don't use memory or streaming)
```
let's not have a todo, refer to jira ticket ARKQB-137 which has a discussion on streaming for agents as tools

dwmkerr commented on ark/internal/genai/context_utils.go:
```diff
@@ -10,6 +10,11 @@ const (
 	queryIDKey   contextKey = "queryId"
 	sessionIDKey contextKey = "sessionId"
 	queryNameKey contextKey = "queryName"
+	// Execution metadata keys for streaming
```
with the ceomment 'set back with streaming chunks in the 'ark' metadata field, which allows callers to differentiate the source of chunks, such as specific agents in a query against at team

dwmkerr commented on ark/internal/genai/context_utils.go:
```diff
@@ -36,3 +41,42 @@ func getSessionID(ctx context.Context) string {
+
+// WithExecutionMetadata adds execution metadata to context for streaming
+func WithExecutionMetadata(ctx context.Context, metadata map[string]interface{}) context.Context {
+	// Avoid nested context in loop by accumulating in temporary variable
```
no breadcrumbs and is there a cleaner way without a nolint

dwmkerr commented on ark/internal/genai/memory_http.go.bak:
```diff
@@ -5,11 +5,14 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io"
```
remove this bak file

dwmkerr commented on ark/internal/genai/memory_http_streaming.go:
```diff
@@ -0,0 +1,471 @@
+}
+
+// HTTPMemoryAndStreaming handles both memory operations and streaming for ARK queries
+type HTTPMemoryAndStreaming struct {
```
there should not be a struct like this there is http memory and http event stream, let's discuss.

dwmkerr commented on ark/internal/genai/memory_http_streaming.go:
```diff
@@ -0,0 +1,471 @@
+	streamMutex  sync.Mutex
+}
+
+func NewHTTPMemory(ctx context.Context, k8sClient client.Client, memoryName, namespace string, recorder EventEmitter, config Config) (MemoryInterface, error) {
```
http memory should be in http_memory and similar for http_streaming

dwmkerr commented on ark/internal/genai/model_generic.go:
```diff
@@ -44,13 +55,25 @@ func (m *Model) ChatCompletion(ctx context.Context, messages []Message, tools []
+
+	// Use streaming if event stream is provided
+	if eventStream != nil {
+		logf.Log.Info("Using streaming mode for chat completion")
```
no noisy logs

dwmkerr commented on ark/internal/genai/model_generic.go:
```diff
@@ -44,13 +55,25 @@ func (m *Model) ChatCompletion(ctx context.Context, messages []Message, tools []
+			return eventStream.StreamChunk(ctx, chunkWithMeta)
+		}, tools...)
+		if response != nil && len(response.Choices) > 0 {
+			logf.Log.Info("Streaming response received",
```
no noisy loggign

dwmkerr commented on ark/internal/genai/model_generic.go:
```diff
@@ -65,3 +88,37 @@ func (m *Model) ChatCompletion(ctx context.Context, messages []Message, tools []
 }
+
+// wrapChunkWithMetadata adds ARK metadata to a streaming chunk
+func (m *Model) wrapChunkWithMetadata(ctx context.Context, chunk *openai.ChatCompletionChunk) interface{} {
```
this should be in streaming.go

dwmkerr replied:
or at least the bulk of the metadata should be and the metadata should be a struct too for clarity, can we then enrich the struct as we enrich context? discuss.

dwmkerr commented on ark/internal/genai/team.go:
```diff
@@ -22,18 +22,25 @@ type Team struct {
 	Recorder    EventEmitter
 	Client      client.Client
 	Namespace   string
+	// Temporary storage for streaming context during execution
```
this comment is not clear i don't think we need it

dwmkerr commented on ark/internal/genai/team.go:
```diff
@@ -206,6 +213,12 @@ func (t *Team) executeWithTracking(tracker *OperationTracker, execFunc func(cont
 func (t *Team) executeMemberAndAccumulate(ctx context.Context, member TeamMember, userInput Message, messages, newMessages *[]Message, turn int) error {
+	// Add team and current member to execution metadata for streaming
+	ctx = WithExecutionMetadata(ctx, map[string]interface{}{
+		"team":  t.Name,
```
we could avoid the hard coding and use teamKey or similar right? across all context usage?

dwmkerr replied:
keys could be in the context file

dwmkerr commented on ark/internal/genai/team.go:
```diff
@@ -215,7 +228,8 @@ func (t *Team) executeMemberAndAccumulate(ctx context.Context, member TeamMember
 	})
 
-	memberNewMessages, err := member.Execute(ctx, userInput, *messages)
+	// Pass through streaming parameters to enable team member streaming
```
no need for this comment

dwmkerr commented on docs/content/developer-guide/ark-sdk.mdx:
```diff
@@ -64,6 +64,21 @@ npm run generate:api
 
+## Adding Custom SDK Functionality
+
+The ARK SDK supports overlay files for adding custom functionality beyond generated types:
```
simplify this, no bold, briefly explain what part of the sdk is generated and what overlay is

dwmkerr commented on lib/ark-sdk/gen_sdk/overlay/python/ark_sdk/streaming_config.py:
```diff
@@ -0,0 +1,97 @@
+
+
+@dataclass
+class ServiceRef:
```
do we use a type like this elsewhere?

dwmkerr commented on lib/ark-sdk/gen_sdk/overlay/python/ark_sdk/streaming_config.py:
```diff
@@ -0,0 +1,97 @@
+
+@dataclass
+class ArkStreamingConfig:
+    """ARK streaming configuration."""
```
note that this is stored in the `ark-config-streaming` config map

dwmkerr commented on lib/ark-sdk/gen_sdk/overlay/python/ark_sdk/streaming_config.py:
```diff
@@ -0,0 +1,97 @@
+    Raises:
+        ValueError: If URL cannot be constructed
+    """
+    if not config:
```
isn't the config non optional anyway? arent't the fields already provided

dwmkerr commented on lib/ark-sdk/gen_sdk/overlay/python/ark_sdk/test_streaming_config.py:
```diff
@@ -0,0 +1,149 @@
+    mock_client = AsyncMock()
+    mock_service = MagicMock()
+    mock_port = MagicMock()
+    mock_port.name = "grpc"
```
we don't use gprc we don't need this

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -165,7 +109,7 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
 
     try:
-        async with with_ark_client("default", "v1alpha1") as ark_client:
+        async with with_ark_client(namespace, "v1alpha1") as ark_client:
```
is v1alpha something we can extract into a constant i think it is used in many place

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -177,39 +121,41 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
-            sse_lines = create_single_chunk_sse_response(completion)
+            # Streaming is enabled - get the base URL and construct full URL
+            base_url = await get_streaming_base_url(streaming_config, namespace, v1)
+            # Add query parameters for OpenAI streaming
```
we need to explain what these parameters re

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -177,39 +121,41 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
+                    iter(sse_lines),
                     media_type="text/event-stream",
-                    headers=sse_headers
+                    headers={
```
Define these earlier in the function so we can use `sseHeaders` and when we define them clearly state what they are for

dwmkerr commented on services/ark-cluster-memory/chart/templates/configmap-streaming.yaml:
```diff
@@ -0,0 +1,12 @@
+{{- if .Values.streaming.enabled }}
+apiVersion: v1
```
i think better to have a clear comment for this and what it is for

dwmkerr commented on services/ark-cluster-memory/chart/values.yaml:
```diff
@@ -38,14 +38,11 @@ memory:
+
+# Streaming configuration
+streaming:
+  # Enable streaming support via ConfigMap
```
via the 'ark-config-streaming' configmap [if i got the name right]

dwmkerr commented on services/ark-cluster-memory/chart/values.yaml:
```diff
@@ -38,14 +38,11 @@ memory:
+  enabled: true
 
 # Persistence configuration
 persistence:
```
there is a `fark` binary below - remove it, find where this file is being created, it should never be in `services`

dwmkerr reviewed (commented):
For review

dwmkerr commented:
Working through the review comments:
1. ✅ Removed .bak file 
2. ✅ Removed breadcrumb comment at line 724 ("Execute without streaming (existing logic)")
3. ✅ Cleaned up a2a_types.go - removed unused A2ASkill type alias and misleading legacy types comment
4. 🔄 Next: Moving stream closing logic from defer to end of query resolution

dwmkerr committed:
refactor: address initial PR review feedback

- Remove unused A2ASkill type alias and misleading legacy comment
- Move event stream closing from defer to explicit close after NotifyCompletion
- Remove breadcrumb comment about existing logic
- Keep stream close errors as logged warnings without failing query

dwmkerr commented:
Re: Stream closing error handling - I've moved the close to after NotifyCompletion as suggested. Keeping close errors as logged warnings rather than failing the query since:
- The query has already completed successfully at this point
- Stream closing is cleanup, not core functionality  
- Failing an otherwise successful query due to cleanup issues seems too harsh
Let me know if you'd prefer different behavior here.

dwmkerr commented:
Re: IsStreamingEnabled helper function - Created genai/query.go with IsStreamingEnabled(query) helper function as suggested. Kept eventStream as a parameter (not in context) for consistency with memory parameter.

dwmkerr committed:
refactor: add IsStreamingEnabled helper function

- Create genai/query.go with IsStreamingEnabled helper
- Replace direct annotation checks with helper function
- Remove unused annotations import from query_controller
- Keep eventStream as parameter for consistency with memory

dwmkerr committed:
refactor: clean up CompleteWithTokens API and add clarifying comments

- Remove unused result parameter from CompleteWithTokens
- Add comment explaining token usage tracking in streaming execution
- Add comment about agent metadata for streaming visibility
- Simplify all CompleteWithTokens callers

dwmkerr committed:
refactor: remove debug logging from agent.go

- Remove excessive info logging for streaming tool calls
- Keep only error logging for tool execution failures
- Simplify control flow after removing debug statements

dwmkerr committed:
refactor: remove noisy logging and unnecessary comments

- Remove debug logging from model_generic.go streaming
- Remove unclear comment about temporary storage in team.go
- Remove redundant comment about streaming parameters
- Replace TODO with JIRA ticket reference ARKQB-137

dwmkerr committed:
refactor: create StreamMetadata struct and move wrapping to streaming.go

- Create proper StreamMetadata struct instead of map[string]interface{}
- Move WrapChunkWithMetadata function to streaming.go
- Remove old wrapChunkWithMetadata method from model_generic.go
- Addresses reviewer feedback about metadata structure and organization

dwmkerr committed:
test: add unit tests for WrapChunkWithMetadata function

- Add comprehensive tests for metadata wrapping with various contexts
- Fix unused import and duplicate type declaration
- Format code with gofmt
- Switch if-else chain to switch statement in tests

dwmkerr committed:
fix: resolve all linting issues

- Add TrueString constant to avoid string literal repetition
- Extract finalizeEventStream method to reduce nesting complexity
- Add nolint comment for test structure consistency
- Remove unnecessary logr import

dwmkerr committed:
docs: add comment explaining streaming metadata purpose

- Clarify that metadata is sent back in 'ark' field with streaming chunks
- Allows callers to differentiate chunk sources in multi-agent queries

dwmkerr committed:
docs: improve commenting

dwmkerr committed:
fix: use query.Spec.SessionId instead of empty string

The SessionId is already available in the QuerySpec and should be used
for the streaming context instead of an empty string.

dwmkerr committed:
fix: remove incorrectly committed services/fark directory and improve documentation

- Remove services/fark directory that was accidentally re-introduced in db20d425
  (fark was correctly moved to tools/fark in c22607cb)
- Remove noisy agent execution log from query controller
- Simplify ark-sdk overlay documentation
- Improve streaming configuration documentation and comments
- Fix unnecessary validation checks in streaming_config.py
- Add clear SSE headers documentation in openai.py

dwmkerr committed:
Merge branch 'main' from upstream

dwmkerr committed:
fix: correct typo 'recieve' to 'receive' in comment

skokaina commented on ark/internal/genai/operation_tracker.go:
```diff
@@ -61,11 +61,7 @@ func (t *OperationTracker) CompleteWithMetadata(result string, additionalMetadat
 }
 
-func (t *OperationTracker) CompleteWithTokens(result string, tokenUsage TokenUsage) {
-	log := logf.FromContext(t.ctx)
```
why are we removing this log ?

dwmkerr replied:
FYI this param had never been used and was set to `""` for callers so wasn't needed

dwmkerr committed:
Merge branch 'main' into integration/streaming

dwmkerr committed:
fix: extend assert timeout for structured output test

Model probes can take time to validate availability, especially with external endpoints

dwmkerr committed:
fix: merge upstream main and resolve RBAC conflicts

- Keep streaming-related permissions (configmaps, services)
- Follow upstream's namespace-scoped Role pattern (no ClusterRole)
- Include Gateway API permissions from upstream
- Consolidate all ark resources permissions in the Role

skokaina committed:
fix: missing responseFormat is openAPI params

skokaina committed:
fix: build issue
